### PR TITLE
ECOM-6051 instructor module

### PR DESCRIPTION
**Pull request by tasawernawaz**
[ECOM-6051](https://openedx.atlassian.net/browse/ECOM-6051)

* Add typeahead search for instructors
* Add Instructor module on new course creation page

**Dropdown:**
screen shot 2017-01-20 at 6 08 54 pm

**Selection:**
screen shot 2017-01-20 at 5 55 33 pm

### CODE REVIEW

**Comment by tasawernawaz** [comment]
@waheedahmed Please review.

File: course_discovery/static/sass/publisher/publisher.scss [hunk 1]
```diff
@@ -814,3 +814,46 @@ select {
...
+  width: 500px;
+}
+
+.instructor-option {
+  img {
+    width: 80px;
+    height: 80px;
+  }
+  td:nth-child(2) {
+    padding-left: 5px;
```
Comment (waheedahmed) [inline course_discovery/static/sass/publisher/publisher.scss]: Add RTL support. ```@include padding-left(5px);```

File: course_discovery/static/sass/publisher/publisher.scss [hunk 2]
```diff
@@ -814,3 +814,46 @@ select {
...
+    padding-right: 5px;
```
Comment (waheedahmed) [inline course_discovery/static/sass/publisher/publisher.scss]: Add RTL support. 
```
@include padding-left(5px);
@include padding-right(5px);
```

File: course_discovery/static/sass/publisher/publisher.scss [hunk 3]
```diff
@@ -814,3 +814,46 @@ select {
...
+  }
+}
+
+.instructor {
+  font-size: smaller;
+  width: 125px;
+  float: left;
```
Comment (waheedahmed) [inline course_discovery/static/sass/publisher/publisher.scss]: Add RTL support. ```@include float(left);```

File: course_discovery/templates/publisher/add_course_form.html [hunk 4]
```diff
@@ -316,17 +316,20 @@ <h2 class="layout-title">{% trans "About page information" %}</h2>
...
                     </div>
-                    <div class="col col-6">
+                    <div class="col col-6 instructor-select">
+                        <label class="field-label ">{{ run_form.staff.label_tag }}</label>
                         {{ run_form.staff }}
                         {% if publisher_add_instructor_feature %}
                             <button type="button" id="add-new-instructor">{% trans "Add New" %}<br>{% trans "Instructor" %}</button>
                         {% endif %}
                     </div>
+                  <div class="selected-instructor">
```
Comment (waheedahmed) [inline course_discovery/templates/publisher/add_course_form.html]: With this selected instructor's will not appear at right place, move this div inside ```<div class="col col-6 instructor-select">```.
screen shot 2017-01-20 at 6 30 26 pm

Comment (tasawernawaz) [reply course_discovery/templates/publisher/add_course_form.html]: this was issue due to rebase, fixed now.

**Comment by tasawernawaz** [comment]
@waheedahmed Your suggestions are implemented. Please review again.

**Review by waheedahmed — APPROVED** [review]
Looks great, nice work!

**Comment by tasawernawaz** [comment]
@bderusha Please review.

File: course_discovery/apps/course_metadata/lookups.py [hunk 5]
```diff
@@ -51,3 +51,31 @@ def get_queryset(self):
...
+            qs = Person.objects.all()
+            if self.q:
+                qs = qs.filter(Q(given_name__icontains=self.q) | Q(family_name__icontains=self.q))
+
+            return qs
+
+        return []
+
+    def get_result_label(self, result):
+        html_start = '''
```
Comment (bderusha) [inline course_discovery/apps/course_metadata/lookups.py]: No HTML in the views.  Please put all of this into a template and return the rendered template

File: course_discovery/apps/publisher/forms.py [hunk 6]
```diff
@@ -18,6 +18,12 @@ def label_from_instance(self, obj):
         return obj.get_full_name()
 
 
+class InstructorModelMultipleChoiceField(forms.ModelMultipleChoiceField):
+    def label_from_instance(self, obj):
+        return '<img src="{profile_image}"/><span>{full_name}</span>'.format(profile_image=obj.profile_image_url,
```
Comment (bderusha) [inline course_discovery/apps/publisher/forms.py]: Same here.  Please no HTML in the view.  Extract this out into it's own small template.

File: course_discovery/apps/course_metadata/urls.py [hunk 7]
```diff
@@ -5,13 +5,14 @@
...
+    CourseAutocomplete, CourseRunAutocomplete, OrganizationAutocomplete, VideoAutocomplete,
+    InstructorAutoComplete)
 
 urlpatterns = [
     url(r'^update_course_runs/(?P<pk>\d+)/$', CourseRunSelectionAdmin.as_view(), name='update_course_runs',),
     url(r'^course-autocomplete/$', CourseAutocomplete.as_view(), name='course-autocomplete',),
     url(r'^course-run-autocomplete/$', CourseRunAutocomplete.as_view(), name='course-run-autocomplete',),
     url(r'^organisation-autocomplete/$', OrganizationAutocomplete.as_view(), name='organisation-autocomplete',),
     url(r'^video-autocomplete/$', VideoAutocomplete.as_view(), name='video-autocomplete',),
+    url(r'^instructor-autocomplete/$', InstructorAutoComplete.as_view(), name='instructor-autocomplete',),
```
Comment (bderusha) [inline course_discovery/apps/course_metadata/urls.py]: `person-autocomplete`

File: course_discovery/apps/course_metadata/lookups.py [hunk 8]
```diff
@@ -51,3 +51,31 @@ def get_queryset(self):
             return qs
 
         return []
+
+
+class InstructorAutoComplete(autocomplete.Select2QuerySetView):
```
Comment (bderusha) [inline course_discovery/apps/course_metadata/lookups.py]: Let's use the generic `PersonAutocomplete` (note the consistent lowercase 'c' in Autocomplete as well)

File: course_discovery/apps/publisher/forms.py [hunk 9]
```diff
@@ -18,6 +18,12 @@ def label_from_instance(self, obj):
         return obj.get_full_name()
 
 
+class InstructorModelMultipleChoiceField(forms.ModelMultipleChoiceField):
```
Comment (bderusha) [inline course_discovery/apps/publisher/forms.py]: `PersonModelMultipleChoice`

File: course_discovery/static/js/publisher/publisher.js [hunk 10]
```diff
@@ -94,3 +103,32 @@ function closeModal(event, modal) {
...
+    e.preventDefault();
+    var id = this.id,
+        option = $('#id_staff').find('option[value="' + id + '"]');
+
+    option.prop("selected", false);
+    this.closest('.selected-instructor, .instructor').remove();
+});
+
+function renderSelectedInstructor(id, name, image) {
+    $('.selected-instructor').append('<div class="instructor"><div><img src="' + image + '"></div><div><a id="' +
```
Comment (bderusha) [inline course_discovery/static/js/publisher/publisher.js]: NIT:  pull HTML elements into a well named variable append that variable for better readability

File: course_discovery/static/sass/publisher/publisher.scss [hunk 11]
```diff
@@ -814,3 +814,46 @@ select {
...
+    text-align: center;
+  }
+  a {
+    color: red;
+  }
+}
+
+.instructor-select {
+  .select2-selection__choice, .select2-selection__clear{
+    display: none!important;
```
Comment (bderusha) [inline course_discovery/static/sass/publisher/publisher.scss]: I am wary of using `!important` to override CSS.  Why is this necessary?  Is there another way to achieve the same effect?

**Review by bderusha — CHANGES_REQUESTED** [review]

**Comment by tasawernawaz** [comment]
@bderusha Feedback addressed, Please review again.

File: course_discovery/apps/course_metadata/lookups.py [hunk 12]
```diff
@@ -51,3 +52,25 @@ def get_queryset(self):
...
+            qs = Person.objects.all()
+            if self.q:
+                qs = qs.filter(Q(given_name__icontains=self.q) | Q(family_name__icontains=self.q))
+
+            return qs
+
+        return []
+
+    def get_result_label(self, result):
+        ctx = {
```
Comment (bderusha) [inline course_discovery/apps/course_metadata/lookups.py]: NIT: full var name please `context`

File: course_discovery/apps/publisher/forms.py [hunk 13]
```diff
@@ -19,6 +20,15 @@ def label_from_instance(self, obj):
         return obj.get_full_name()
 
 
+class PersonModelMultipleChoice(forms.ModelMultipleChoiceField):
+    def label_from_instance(self, obj):
+        ctx = {
```
Comment (bderusha) [inline course_discovery/apps/publisher/forms.py]: NIT: full var name `context`

**Review by bderusha — APPROVED** [review]

**Pull request merged**